File: lista/GO/questao12/questao12.go
package main

import (
	"fmt"
)

type Produto struct {
	Nome  string
	Preco float64
}

func (p Produto) Somar(outro Produto) float64 {
	return p.Preco + outro.Preco
}

func main() {
	produto1 := Produto{
		Nome:  "Livro",
		Preco: 25.50,
	}

	produto2 := Produto{
		Nome:  "Caderno",
		Preco: 40.00,
	}

	precoTotal := produto1.Somar(produto2)

	fmt.Printf("Preço total de %s e %s é: R$ %.2f\n", produto1.Nome, produto2.Nome, precoTotal)
}

File: lista/Python/questao12.py
"""
Em Python, sobrecarregue o operador + para somar dois objetos Produto
baseados no preço.
"""
class Produto:
    def __init__(self, nome, preco):
        self.nome = nome
        self.preco = preco

    def __add__(self, other):
        if isinstance(other, Produto):
            return Produto(f"{self.nome} e {other.nome}", self.preco + other.preco)
        return NotImplemented

    def __str__(self):
        return f"Produto: {self.nome}, Preço: R${self.preco:.2f}"

def main():
    produto1 = Produto("Livro", 25.90)
    produto2 = Produto("Caderno", 40.90)

    produto3 = produto1 + produto2

    print(produto1)
    print(produto2)
    print(produto3)

if __name__ == "__main__":
    main()
